cpp/include/rapidsmpf/memory/pinned_memory_resource.hpp

```diff
@@ -242,184 +243,4 @@ class PinnedMemoryResource {
- *
- * \sa
- * https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY__POOLS.html#group__CUDART__MEMORY__POOLS_1g871003f518e27ec92f7b331307fa32d4
- *
- * @code{.cpp}
- * rapidsmpf::PinnedHostBuffer buffer1(1024, stream, mr);
- * rapidsmpf::PinnedHostBuffer buffer2(1024, stream, mr);
- *
- * // data() is immediately available for cuda*Async operations. Eg.
- * cudaMemcpyAsync(buffer2.data(), buffer1.data(), 1024, cudaMemcpyDefault, stream);
- *
- * // for stream-unaware operations, the stream must be synchronized.
- * std::vector<uint8_t> data(1024);
- *
- * // synchronize buffer1
- * buffer1.synchronize();
- *
- * // now buffer1 is ready to use in stream-unaware operations.
- * std::memcpy(data.data(), buffer1.data(), 1024);
- * std::cout << buffer1.data()[0] << ", " << data[0] << std::endl;
- * @endcode
- */
-class PinnedHostBuffer {
-  public:
-    PinnedHostBuffer() = default;
-
-    /**
-     * @brief Constructs an empty pinned host buffer.
-     *
-     * @param size The size of the buffer in bytes.
-     * @param stream The CUDA stream to use for memory operations.
-     * @param mr Memory resource to use for allocation and deallocation.
-     *
-     * @throws std::invalid_argument If @p mr is nullptr.
-     */
-    PinnedHostBuffer(
-        size_t size,
-        rmm::cuda_stream_view stream,
-        std::shared_ptr<PinnedMemoryResource> mr
-    );
-
-    /**
-     * @brief Constructs a pinned host buffer and asynchronously copies data into it.
-     *
-     * @param src_data Pointer to the source data to copy. If the data ptr is ordered on a
-     * stream other than @p stream, it must be synchronized with @p stream before calling
-     * this constructor. Otherwise, it will result in undefined behavior.
-     * @param size The size of the data to copy in bytes.
-     * @param stream The CUDA stream to use for memory operations.
-     * @param mr Memory resource to use for allocation and deallocation.
-     *
-     * @throws std::invalid_argument If @p src_data is nullptr or @p data_ is nullptr (ie.
-     * allocation failed).
-     * @throws rapidsmpf::cuda_error If @p cudaMemcpyAsync fails.
-     */
-    PinnedHostBuffer(
-        void const* src_data,
-        size_t size,
-        rmm::cuda_stream_view stream,
-        std::shared_ptr<PinnedMemoryResource> mr
-    );
-
-    /**
-     * @brief Constructs a pinned host buffer by copying data asynchronously from another
-     * buffer on the same stream as @p other.
-     *
-     * @tparam OtherBufferT The type of the other buffer to copy from. Eg.
-     * `rmm::device_buffer` or `PinnedHostBuffer`.
-     *
-     * @param other The other pinned host buffer to copy from.
-     * @param mr Memory resource to use for allocation and deallocation.
-     */
-    template <typename OtherBufferT>
-    PinnedHostBuffer(OtherBufferT const& other, std::shared_ptr<PinnedMemoryResource> mr)
-        : PinnedHostBuffer(other.data(), other.size(), other.stream(), std::move(mr)) {}
```

Comment:
@madsbk IMO we should port this constructor. It would be easy to create `HostBuffer` from any buffers here.